proposals/0000-type-vs-constraint.rst

```diff
@@ -0,0 +1,393 @@
+whereas the kind of ``a`` is ``Type``.
+
+GHC allows us to extract out an equality relationship between *kinds* from an
+equality relationship on *types* -- and kind equalities are always nominal. To
+wit, Core allows ::
+
+  KindCo axDefault :: Constraint ~# Type
+
+Now, suppose that you could write this::
+
+  type family F a
+  type instance F Type = Int
+  type instance F Constraint = Bool
+
+If these instances were allowed, GHC could
+produce a coercion between ``Int`` and ``Bool``, thus::
+
+  Bool  ~#  F Constraint   -- By type instance F Constraint (backwards)
+        ~#  F Type         -- By KindCo axDefault
+        ~#  Int            -- By type instance F Type
+
+That would be Very, Very Bad.  So, although ``Type`` and ``Constraint`` are built
+with different (un-equal) primitive type constructors,
+
+* **GHC's type checker treats `Type` and `Constraint` as not "apart".**
+
+That in turn makes GHC complain that the above instances overlap, and are hence illegal.
+You can read more about what "apart" means in
+`Closed type families with overlapping equations <https://simon.peytonjones.org/closed-type-families/>`_,
+sections 3.2 and 3.3.
+
+Apartness affects type-class instances as well as type-family instances.
+Suppose we have::
+
+  instance {-# OVERLAPPABLE #-} C Int a  where ...
+  instance {-# OVERLAPPING #-}  C Int Bool where ...
+
+and we are trying to solve the constraint ``C Int (F gamma)``, where ``F`` is a type family and ``gamma``
+is an as-yet-unknown unification variable. Since ``(F gamma)`` is not
+apart from ``Bool``, it could be that (once we know ``gamma``) the second
+instance should be chosen. So GHC declines to commit to the first.
+
+If ``(F gamma)`` later simplifes to, say ``Char`` (which *is* apart from ``Bool``),
+then and only then GHC can commit to the first instance.
+
+Similarly with ``Type`` and ``Constraint``. Suppose we have::
+
+  instance D Type where...
+  instance D Constraint where ...
+
+and try to solve ``C Int Type``. This matches the first instance, but
+since ``Type`` is not apart from ``Constraint`` GHC thinks "oh, the
+second one could match in the future" and declines to commit to the
+first.  In fact, you can have either of these instances separately,
+but if they occur together neither will ever be chosen: they overlap
+irretrievably.
+
+
+Proposed Change Specification
+-----------------------------
+
+We propose the following new setup, not repeating any types that remains unchanged::
+
+  -- Primitive type constructors
+  type SORT :: TypeOrConstraint -> RuntimeRep -> Type
+  type IP   :: forall (r :: RuntimeRep). Symbol -> TYPE r -> CONSTRAINT r
+
+  type (=>)  :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> TYPE r2 -> Type  -- primitive
+  type (==>) :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> CONSTRAINT r2 -> Constraint
+
+  -- Data types
+  data TypeOrConstraint = TypeLike | ConstraintLike
```

Comment:
I don't think this would affect the proposal in a fundamental way, particularly since TypeOrConstraint will be an internal detail, but do we expect to ever add new cases here, e.g. as part of Kinds are Calling Conventions?

Perhaps another way of asking this question is: you cite Kinds are Calling Conventions (KCC) as an extension this proposal would better position us to support in the future, can you quickly sketch how you adapt the proposal to support KCC? Not being super familiar with that work it's hard for me to judge how much the current proposal buys us on that front.

Reply:
Adapting [Kind as Calling Conventions](https://www.microsoft.com/en-us/research/publication/kinds-are-calling-conventions/) to this proposal we might get:
```
type SORT :: Convention -> TypeOrConstraint -> RuntimeRep -> Type
data Convention = Eval levity | Call ArityDescription
data ArityDecription = ACons RuntimeRep ArityDescription | AZero | AConv Convention
```
Here we add a new `Convention` argument to `SORT` that describes the arity of the thing.

I'm not certain that this is the right thing to do, but it's plausible.

My long term plan is this: 
* in Core we make sure that the argument of a function has the *type* expected by the function
* in STG we make sure that the argument of a function has the *kind* expected by the function

So in STG we have a coarser well-typed-ness condition -- but that coarsening in turn allows us to do more CSE and unboxed sums and other such things.

But to call functions with the right calling convention we would really need to express the calling convention in the kind.  Hence KACC.